### ISSUE TITLE
NAME hangs in tight loop

### ISSUE THREAD

**Issue by eswenson1** [PDP-10/its#1505]
After updating the host table with :SYSHST;H3MAKE BIN, I got into a situation where NAME hangs. It is in a tight loop here:

```
gotda1:  move u,b
        skipl b,@suppro
         jrst gotda1            ;loop till find -1, which means job is top level
```

This code appears wrong.  Nothing will ever update b, so if we ever get to gotda1, we'll never get out of this loop.  I traced through NAME, and don't understand what is going on.  It appears that we are iterating over all the jobs, but the data at the first job we find (that is associated with a TTY) doesn't appear to be an entry in the JOB table.  The UNAME/JNAME are garbage.

**Comment by eswenson1** [comment]
When it gets into this state, the way to fix is to put a breakpoint at GOC and patch the value returned by .RSYSID to something other than the current ITS version (0 works).  This will force NAME to reinitialize and repurify itself.  It automatically PDUMPs itself.  

I'm not sure why it needs to be reinitialized.  I didn't see anything wrong with the host table produced by H3MAKE and all host lookups seem to be working fine.  Just NAME misbehaves.

**Comment by larsbrinkhoff** [comment]
I think looks are deceiving here. The move from @SUPPRO accesses an internal ITS data structure.  Note that SUPPRO is actually indexed off U.  On the next iteration, U is updated from B.  So this walks the job tree until there is a negative value which indicates we reached a toplevel job.

**Comment by larsbrinkhoff** [comment]
Sorry @suppro, you're named after a symbol in ITS.

**Comment by eswenson1** [comment]
Yeah, I kind of figured that out because the same thing is done with TTYSTS.  However, I don't understand HOW this is done.  Doesn't an instruction like MOVE A,@SUPPRO just look in SUPPRO for an address and then load from that address?  Where is the U indexing happening?  How does either the MOVE instruction know to index off of U, or how does SUPPRO encode this such that the hardware will do the right thing?  Isn't SUPPRO just an .eval'ed symbol in ITS?  Where is the U offset factored in?

**Comment by larsbrinkhoff** [comment]
@ does indirection, so there's a new effective address calculation done with the word at suppro.  I get:

    SUPPRO/ U,,554356

Look how the call to .EVAL involves an argument, "get addr to be iored into what eval gives"

```
        pushj p,evaler
         400000(u)
          usrxtb

**Comment by eswenson1** [comment]
Ahhh, I didn't know you could do that!  Very cool.  So the indirection is in the contents of the SUPPRO address in NAME's address space. I saw ITS doing a SUPPRO(U) all over the place, but didn't see NAME doing it -- didn't realize it was turning the address of SUPPRO in ITS into one with automatic indexing off of U.  Again, very cool.

**Comment by eswenson1** [comment]
But still: NAME does hang in this case.  And it must be that SUPPRO is pointing to the wrong place, because it never exits the loop. U was never changing, so we were always looking at the same SUPPRO(U) address, and never getting anywhere. It goes on indefinitely -- which means that SUPPRO or U were incorrect to begin with.  SInce re-purifying NAME "fixes" this, I'm wondering what "broke" it.

**Comment by larsbrinkhoff** [comment]
This points to ITS having been changed after NAME was purified the first time.  E.g. assembling a new ITS without updating the version number will do it.  It happens to me all the time, but I usually see it with PEEK.

**Comment by eswenson1** [comment]
Yes, and it is probably the case that all this is a big coincidence. I probably never repurified NAME after I updated ITS to add the chaosnet config.  However, I jumped to a conclusion because you reported that when you used H3MAKE to rebuild the host table, NAME hung.  Since I had just used H3MAKE to build my host table, I assumed I ran into the same thing you did.

So my hang makes sense -- I probably needed to repurify NAME.  But H3MAKE works perfectly fine for me in DB/EX as well as ES.  So I'm wondering, now, why it caused NAME to hang for you!  :-)

**Comment by larsbrinkhoff** [comment]
I ran it like this: `:syshst;h3make` which uses a link to H3MAKE BIN.

I tested it on two machines, and it seemed fine.  Got favourably output in H3TYO.

This could be debugged by doing a SYSNET^F in the TCL script right before NAME is purified.  If there's a H3ERR there, something went wrong.
